### DISCUSSION TITLE
Unable to add Intel RealSense Camera device

### DISCUSSION THREAD

**Discussion by Roopesh-Nallakshyam** [PlusToolkit/PlusLib#949] [Q&A]
I successfully built PlusApp in Linux Ubuntu 20.04.4, by following build steps given here https://github.com/PlusToolkit/PlusBuild/blob/master/Docs/BuildInstructionsLinux.md. When I launch PlusServerLaunch binary it is showing me in terminal , _**Supported devices: ChRobotics, FakeTracker, GenericSerialDevice, ImageProcessor, Microchip, NoiseVideo, OpenIGTLinkTracker, OpenIGTLinkVideo, SavedDataSource, UsSimulator, VirtualBufferedCapture, VirtualCapture, VirtualDeinterlacer, VirtualDiscCapture, VirtualMixer, VirtualSwitcher, VirtualVolumeReconstructor**_

* But I am using Intel Realsense D435i camera, which is not listed above. Then I looked at page to add new device https://plustoolkit.github.io/devicecode , followed steps as mentioned. Created one [External_IntelRealSense_Device .cmake.txt](https://github.com/PlusToolkit/PlusLib/files/8725081/External_IntelRealSense_Device.cmake.txt).
* Started building PlusApp with new Intel RealSense device , then I am getting gowlling error at vtkPlusDataCollection.cxx

`/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_config_disable_all_streams'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_process_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_pipeline_start_with_config'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_stream_profiles_list'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_failed_function'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_pipeline'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_query_sensors'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_context'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_stream_profile_data'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_frame_queue'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_is_sensor_extendable_to'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_pipeline_profile_get_streams'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_context'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_is_stream_profile_default'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_start_processing'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_pipeline_wait_for_frames'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_sensor'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_failed_args'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_depth_scale'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_stream_profile'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_frame_queue'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_device'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_frame_stream_profile'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_stream_profiles_count'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_free_error'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_poll_for_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_error_message'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_embedded_frames_count'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_release_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_frame_add_ref'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_pipeline'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_processing_block'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_enqueue_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_sensor_list'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_sensor'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_config_enable_stream'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_config'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_frame_data'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_is_frame_extendable_to'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_delete_pipeline_profile'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_extract_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_config'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_sensors_count'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_keep_frame'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_get_librealsense_exception_type'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_pipeline_stop'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_colorizer'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_pipeline_profile_get_device'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_create_align'
/usr/bin/ld: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/libvtkPlusDataCollection.so: undefined reference to `rs2_set_option'
collect2: error: ld returned 1 exit status
make[5]: *** [src/PlusDataCollection/Testing/CMakeFiles/TransformInterpolationTest.dir/build.make:276: /home/comofi/Desktop/PlusBuild/Plus_Bin_2_9/bin/TransformInterpolationTest] Error 1
make[4]: *** [CMakeFiles/Makefile2:3239: src/PlusDataCollection/Testing/CMakeFiles/TransformInterpolationTest.dir/all] Error 2
`

Please help me here to solve this?

**Comment by adamrankin** [comment]
I commend you for the effort in adding a new device, but the RealSense is already supported. I will confirm that it's also supported in Linux.

**Reply by adamrankin** [reply]
After a couple of small fixes I am able to build Plus with realsense enabled. Are you able to try it?

**Reply by Roopesh-Nallakshyam** [reply]
No, I am unable to do it, can you explain me ?

**Reply by adamrankin** [reply]
Yes, you'll have to build Plus with realsense enabled. In PlusBuild Cmake, enable the checkbox (if cmake-gui) or ON/OFF (if ccmake) for RealSense and rebuild PlusBuild. It will complain and ask you for the location of your realsense sdk. I found the easiest was to install librealsense2-dev and then give it the library location
